Framework/Core/src/DataProcessingDevice.cxx

```diff
@@ -1460,7 +1460,11 @@ bool DataProcessingDevice::tryDispatchComputation(DataProcessorContext& context,
         continue;
       }
       // in DPL we are using subchannel 0 only
-      device->Send(forwardedParts[fi], spec->forwards[fi].channel, 0);
+      // Notice that we cannot block, because we will not
+      // be unblocked in case of a new state
+      // device->Send(forwardedParts[fi], spec->forwards[fi].channel, 0);
+      while (device->Send(forwardedParts[fi], spec->forwards[fi].channel, 0, 100) == 0 || device->NewStatePending() == false) {
```

Comment:
@rbx this does not do what I want, I guess. How can I make sure that all the parts have been sent?

Reply:
Return value of `0` can be success, if messages are empty. I think for what you want you should use `< 0`, or `<= 0` if you are sure there are no empty messages.
The return values below than 0 are:
    error = -1,
    timeout = -2,
    interrupted = -3
If it is 0 or larger, then it means all parts have been sent out. There is no partial transfer, it's either all or none.

Reply:
Ok, thanks. I adapted the code accordingly. I think I will simply move that responsibility to libuv as well, as already done for the inputs. Do I understand correctly that indeed the state change will not unblock a blocking send?

Reply:
A state change will unblock a blocking Send/Receive call, which will return `TransferCode::interrupted` (-3). After such unblock you can still issue further transfer calls - they will attempt transfer for ~100ms long and if not successful return `TransferCode::interrupted` again. If successful, they will return number of bytes as usual.

Reply:
> A state change will unblock a blocking Send/Receive call

this I do not see happening, not with the static plugin, at least.

Reply:
It does for me. All controllers interrupt in the same way, there should be no difference there.
Try to play with this simple device: https://github.com/rbx/FairMQ/blob/check-sink/fairmq/devices/runSink.cxx
Run it with static/interactive controller:
```
./fairmq/fairmq-sink --id sink1 --channel-config name=data,method=bind,type=pull,address=ipc://i --control static
```